### ISSUE TITLE
In the accessible buffer the whitespace is apparently being replaced by something other than whitespace.

### ISSUE THREAD

**Issue by jvesouza** [microsoft/vscode#181834]
Type: <b>Bug</b>
CC: @meganrogge  @isidorn

I noticed that I can't navigate the accessible buffer using ctrl left arrow or ctrl+right arrow. When using this combination orca treats the whole line as if it were a single word.
Except for misunderstandings, this happens because the blank spaces present in the accessible buffer are being replaced by the character 0xa0. 
I ran the command echo "this is test" in the VSCode terminal and ran the same command in the gnome terminal. Browsing character by character with orca I noticed that in the gnome terminal space appears as 0x20 and in the accessible buffer it appears as 0xa0.

VS Code version: Code - Insiders 1.79.0-insider (2c73d2651cc525b27fc697b820743098a204ec01, 2023-05-04T22:12:16.054Z)
OS version: Linux x64 6.2.14-300.fc38.x86_64
Modes:
Sandboxed: Yes
Remote OS version: Linux x64 4.14.146-93.123.amzn1.x86_64

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz (8 x 2899)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|2, 2, 2|
|Memory (System)|15.26GB (6.34GB free)|
|Process Argv|--crash-reporter-id bc02adbd-6c3a-44b8-af92-55103a608af7|
|Screen Reader|yes|
|VM|0%|
|DESKTOP_SESSION|gnome-xorg|
|XDG_CURRENT_DESKTOP|GNOME|
|XDG_SESSION_DESKTOP|gnome-xorg|
|XDG_SESSION_TYPE|x11|

|Item|Value|
|---|---|
|Remote|SSH: informal|
|OS|Linux x64 4.14.146-93.123.amzn1.x86_64|
|CPUs|Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz (2 x 2500)|
|Memory (System)|3.80GB (1.63GB free)|
|VM|0%|
</details><details><summary>Extensions (44)</summary>

Extension|Author (truncated)|Version
---|---|---
doxdocgen|csc|1.4.0
vscode-markdownlint|Dav|0.50.0
vscode-eslint|dba|2.4.0
EditorConfig|Edi|0.16.4
copilot|Git|1.86.82
rest-client|hum|0.25.1
better-cpp-syntax|jef|1.17.2
docomment|k--|1.0.0
azure-pipelines|ms-|1.208.0
vscode-docker|ms-|1.25.0
csharp|ms-|1.25.7
vscode-dotnet-runtime|ms-|1.6.0
vscode-kubernetes-tools|ms-|1.3.11
remote-containers|ms-|0.293.0
remote-ssh|ms-|0.101.2023050515
remote-ssh-edit|ms-|0.84.0
azure-account|ms-|0.11.4
azurecli|ms-|0.5.0
cmake-tools|ms-|1.14.31
cpptools|ms-|1.15.4
cpptools-extension-pack|ms-|1.3.0
makefile-tools|ms-|0.7.0
powershell|ms-|2023.5.0
remote-explorer|ms-|0.5.2023040509
vsliveshare|ms-|1.0.5864
oracledevtools|Ora|21.5.0
fabric8-analytics|red|0.3.6
java|red|1.19.2023050804
vscode-xml|red|0.25.2023050404
vscode-yaml|red|1.12.2
sonarlint-vscode|Son|3.17.0
cmake|twx|0.0.17
intellicode-api-usage-examples|Vis|0.2.7
vscodeintellicode|Vis|1.2.30
vscode-boot-dev-pack|vmw|0.2.1
vscode-spring-boot|vmw|1.46.0
vscode-java-debug|vsc|0.50.2023042406
vscode-java-dependency|vsc|0.22.2023042900
vscode-java-pack|vsc|0.25.2023042006
vscode-java-test|vsc|0.38.2023042703
vscode-maven|vsc|0.41.2023042103
vscode-spring-boot-dashboard|vsc|0.13.0
vscode-spring-initializr|vsc|0.11.2
gitblame|wad|10.2.0

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256198
vstes627cf:30244335
vslsvsres303:30308271
pythontb:30258533
pythonptprofiler:30281269
vshan820:30294714
pythondataviewer:30285072
vscod805cf:30301675
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593cf:30376535
pythonvs932:30404738
cppdebug:30492333
vsclangdf:30492506
c4g48928:30535728
dsvsc012cf:30540253
pynewext54:30618038
pylantcb52:30590116
pyind779:30611226
pythonsymbol12:30651887
6233i204:30725950
azdwalk:30721579
pythonms35:30671666
cai9j608:30680470
24365598:30687740
pythonfmttext:30716741
pythoncmvfstr:30726892
fixshowwlkth:30724385
hidesbindicator:30724476
pythongtdpath:30726887

```

</details>

**Comment by meganrogge** [comment]
We replace these spaces intentionally, but I'm not sure why 🤔 . @Tyriar thoughts?

https://github.com/microsoft/vscode/blob/a54b497150197f773f485038f44af02772c5a016/src/vs/workbench/contrib/terminalContrib/accessibility/browser/bufferContentTracker.ts#L125

**Comment by jvesouza** [comment]
As it is, I can't use orca to select words and I can't navigate by words either.

**Comment by Tyriar** [comment]
We replace it with a non-breaking space such that white space appears the same as it would in the terminal. I wasn't aware it has this impact on screen readers (maybe it's just orca?) but we could workaround it by replacing `'  '` (2 spaces) with `' \xA0'`. That should keep tables, ls, etc. aligned while fixing the problem at hand. It seems like an Orca issue though as I'm pretty sure a non-breaking space should be getting treated as a space.

**Comment by VSCodeTriageBot** [comment]
This bug has been fixed in the latest release of [VS Code Insiders](https://code.visualstudio.com/insiders/)!

@jvesouza, you can help us out by commenting `/verified` if things are now working as expected.

If things still don't seem right, please ensure you're on version b9a77b3c7b43f229a430b28070250d55e5d14858 of Insiders (today's or later - you can use `Help: About` in the command palette to check), and leave a comment letting us know what isn't working as expected.

Happy Coding!

**Comment by Tyriar** [comment]
This doesn't appear to be fixed:

![image](https://github.com/microsoft/vscode/assets/2193314/d5f9576d-c60d-4070-9856-8f9df442f027)

![image](https://github.com/microsoft/vscode/assets/2193314/2d83e543-d3a8-468d-997e-ea99ff87c76f)

It is changing multiple lines to include a single space character as expected, but see the first line where single spaces are changed. I'm questioning the fix we came up with though, monaco replaces spaces anyway in the editor:

![image](https://github.com/microsoft/vscode/assets/2193314/fe7c8d79-1e57-461c-85bd-96da66827e7c)

I think we should revert this and have orca fix it, instead of trying to workaround their problem which could end up causing other issues for us?

**Comment by jvesouza** [comment]
I ended up reporting this problem to the orca developers because the problem also exists when I'm editing a file that has words separated by the character 0xa0. I believe the problem will be dealt with on the orca side.

**Comment by joanmarie** [comment]
If I understand what @jvesouza reported correctly, I don't *think* this is an Orca issue. 

When Orca is in focus mode, it does not control or consume input events. In other words, Ctrl+Right/Left arrow will do whatever they would do without Orca running. And from some testing, when the words are separated by 0xa0, VS Code (or Chromium) is treating the string as a single word.

If, in those same places, I switch Orca to browse mode, Orca IS controlling/consuming the input events, and it already does treat the string as individual words.

But if I'm missing something, please let me know what I should be fixing in Orca. Thanks!

**Comment by joanmarie** [comment]
@Tyriar:

Here are the steps I am using to reproduce the problem which I believe @jvesouza is reporting:

1. Be sure Orca is NOT running.
2. Create a new document.
3. Type "This"
4. Press Ctrl+Shift+U
5. Type "00A0" followed by return to insert a non-breaking space character
6. Type "is"
7. Repeat steps 4 and 5.
8. Type "a"
9. Repeat steps 4 and 5.
10. Type "test."
11. Press Home to move to the beginning of the line.
12. Press Ctrl+Right to move to the next word.

Expected results: The caret would move just past the "s" of "This".

Actual results: The caret immediately jumps to the "." at the end of the sentence.

**Comment by Tyriar** [comment]
@joanmarie thanks for the follow up, you're right.

@meganrogge we should not be replacing it with non-breaking spaces as monaco handles their presentation. When you paste in the characters like @joanmarie describes (I did it by running '\u00a0' in console and copying that) you can see they're treated specially, we also have a lint rule to prevent this in vscode:

![image](https://github.com/microsoft/vscode/assets/2193314/c84ae292-a6cb-4836-9c55-9ae4f68fe2d5)
